src/irr/video/COpenGLSpecializedShader.cpp

```diff
@@ -114,10 +114,104 @@ COpenGLSpecializedShader::COpenGLSpecializedShader(const video::IVideoDriver* _d
+                {&gl::extGlProgramUniformMatrix4x2fv, &gl::extGlProgramUniformMatrix4x3fv, &gl::extGlProgramUniformMatrix4fv} //4xM
+            };
+            glProgramUniformMatrixNxMfv_fptr[m.mtxColCnt-2u][m.mtxRowCnt-2u](m_GLname, u.location, m.count, GL_FALSE, matrix_data.data());
+        }
+        else if (is_single_or_vec()) {
+            core::vector<GLuint> vector_data(m.count*m.mtxRowCnt);
+            for (uint32_t i = 0u; i < m.count; ++i) {
+                const GLuint* vec = reinterpret_cast<const GLuint*>(_pcData + i*m.mtxRowCnt);
+                vector_data.insert(vector_data.end(), vec, vec+m.mtxRowCnt);
+            }
+            switch (m.type) {
+            case asset::EGVT_F32:
+            {
+                PFNGLPROGRAMUNIFORM1FVPROC glProgramUniformNfv_fptr[4]{
+                    &gl::extGlProgramUniform1fv, &gl::extGlProgramUniform2fv, &gl::extGlProgramUniform3fv, &gl::extGlProgramUniform4fv
+                };
+                glProgramUniformNfv_fptr[m.mtxRowCnt-1u](m_GLname, u.location, m.count, reinterpret_cast<const GLfloat*>(vector_data.data()));
+                break;
+            }
+            case asset::EGVT_I32:
+            {
+                PFNGLPROGRAMUNIFORM1IVPROC glProgramUniformNiv_fptr[4]{
+                    &gl::extGlProgramUniform1iv, &gl::extGlProgramUniform2iv, &gl::extGlProgramUniform3iv, &gl::extGlProgramUniform4iv
+                };
+                glProgramUniformNiv_fptr[m.mtxRowCnt-1u](m_GLname, u.location, m.count, reinterpret_cast<const GLint*>(vector_data.data()));
+                break;
+            }
+            case asset::EGVT_U32:
+            {
+                PFNGLPROGRAMUNIFORM1UIVPROC glProgramUniformNuiv_fptr[4]{
+                    &gl::extGlProgramUniform1uiv, &gl::extGlProgramUniform2uiv, &gl::extGlProgramUniform3uiv, &gl::extGlProgramUniform4uiv
+                };
+                glProgramUniformNuiv_fptr[m.mtxRowCnt-1u](m_GLname, u.location, m.count, vector_data.data());
+                break;
+            }
+            }
+        }
+    }
+}
+
+void COpenGLSpecializedShader::buildUniformsList()
+{
+    const auto& pc = m_introspectionData->pushConstant;
+    if (!pc.present)
+        return;
+
+    const auto& pc_layout = pc.info;
+    core::queue<SMember> q;
+    SMember initial;
+    initial.type = asset::EGVT_UNKNOWN_OR_STRUCT;
+    initial.members = pc_layout.members;
+    q.push(initial);
+    while (!q.empty())
+    {
+        const SMember top = q.front();
+        q.pop();
+        if (top.type == asset::EGVT_UNKNOWN_OR_STRUCT && top.members.count) {
+            for (size_t i = 0ull; i < top.members.count; ++i) {
+                SMember m = top.members.array[i];
+                m.name = top.name + "." + m.name;
+                q.push(m);
+            }
+            continue;
+        }
+        using gl = COpenGLExtensionHandler;
+        const GLint location = gl::extGlGetUniformLocation(m_GLname, top.name.c_str());
```

Comment:
remember that if `top.name` is the name of a variable thats an array you need to query for `top.name+"[0]"`

Reply:
oh, right, thx

Reply:
resolved